### PR TITLE
Pulse gradient methods docs polish

### PR DESCRIPTION
**Pull request by Qottmann**
Polishing some small bits in the docs of the stochastic parameter shift and generator method.

### CODE REVIEW

File: pennylane/gradients/pulse_generator_gradient.py [hunk 1]
```diff
@@ -407,16 +407,15 @@ def _pulse_generator(tape, argnum=None, shots=None, atol=1e-7):
     This method combines automatic differentiation of few-qubit operations with
     hardware-compatible shift rules.
     It allows for the evaluation of parameter-shift gradients for many-qubit pulse programs
-    on hardware, with the limitation that the program must be composed of few-qubit pulses.
+    on hardware, with the limitation that the individual pulses must be acting on a reduced number of qubits.
```
Comment (dwierichs) [inline pennylane/gradients/pulse_generator_gradient.py:410]: Maybe a bit more "concrete"?
```suggestion
    on hardware, with the limitation that the individual pulses must be acting on a 
    sufficiently small number of qubits.
```

File: pennylane/gradients/pulse_generator_gradient.py [hunk 2]
```diff
@@ -407,16 +407,15 @@ def _pulse_generator(tape, argnum=None, shots=None, atol=1e-7):
...
 
-    For this differentiation method, the unitary matrix corresponding to each pulse
-    is computed and differentiated classically with an autodiff framework,
+    For this differentiation method, the unitary matrix :math:`U` and its derivative :math:`\partial_k U`
+    of a pulse gate are computed classically with an autodiff framework. From :math:`\partial_k U` and :math:`U` we can
+    deduce the so-called effective generators :math:`\Omega_{k}` by assuming the form
```
Comment (dwierichs) [inline pennylane/gradients/pulse_generator_gradient.py:414]:
```suggestion
    For this differentiation method, the unitary matrix :math:`U` of a pulse gate and its derivative
    :math:`\partial_k U` are computed classically with an autodiff framework.
    From :math:`\partial_k U` and :math:`U` we can deduce the so-called effective generators
    :math:`\Omega_{k}` by assuming the form
```

File: pennylane/gradients/pulse_generator_gradient.py [hunk 3]
```diff
@@ -619,7 +618,9 @@ def circuit(params):
 
             \omega_{k}^{(\ell)} = \frac{1}{2^N}\mathrm{Tr}\left[P_\ell \Omega_{k}\right]
 
-        where :math:`N` is the number of qubits.
+        where :math:`N` is the number of qubits and :math:`\ell = 1, .. , L` the Pauli word index.
+        The number of non-zero Pauli words :math:`L` is the dimension of the dynamical Lie algebra
+        and at most :math:`4^N-1`.
```
Comment (dwierichs) [inline pennylane/gradients/pulse_generator_gradient.py:623]: I guess this _can_ be incorrect, if we happen to compute the effective generator at a weird position in parameter space, right? :thinking: Maybe inserting a "typically" or "usually" or so could defuse this?

File: pennylane/gradients/pulse_generator_gradient.py [hunk 4]
```diff
@@ -658,14 +659,20 @@ def circuit(params):
         before the parametrized time evolution. Here, the variable :math:`x` is just a
         convenient way to write the modified cost function. Note that its derivative with
         respect to :math:`x` can be computed with the standard two-term parameter-shift
-        rule for Pauli rotation gates.
+        rule for Pauli rotation gates. I.e.
```
Comment (dwierichs) [inline pennylane/gradients/pulse_generator_gradient.py:662]: Personal preference:
```suggestion
        rule for Pauli rotation gates, i.e.
```

File: pennylane/gradients/pulse_gradient.py [hunk 5]
```diff
@@ -412,14 +412,14 @@ def ansatz(params):
...
     >>> qnode(params)
-    Array(-0.0905377, dtype=float32)
+    Array(-0.0905377, dtype=float64)
 
     And as we registered the differentiation method :func:`~.stoch_pulse_grad`,
     we can compute its gradient in a hardware compatible manner:
 
     >>> jax.grad(qnode)(params)
-    [Array(0.00054451, dtype=float32, weak_type=True),
-     Array(-0.05673098, dtype=float32, weak_type=True)] # results may differ
```
Comment (dwierichs) [inline pennylane/gradients/pulse_gradient.py:422]: Did you consciously remove the comment? :)

File: pennylane/gradients/pulse_gradient.py [hunk 6]
```diff
@@ -466,7 +466,7 @@ def ansatz(params):
             times.append(process_time() - start)
 
     >>> print(times) # Show the gradient computation times in seconds.
-    [44.92675887899999, 10.405261402999997]
+    [55.75785480000002, 12.297400500000009]
```
Comment (dwierichs) [inline pennylane/gradients/pulse_gradient.py:469]: :sweat_smile:

File: pennylane/gradients/pulse_gradient.py [hunk 7]
```diff
@@ -542,9 +542,9 @@ def ansatz(params):
 
         .. math::
 
-            H(\boldsymbol{v}, t) = \frac{1}{2} X\otimes I + v_1 Z\otimes Z + \sin(v_2 t) I\otimes X
+            H(\boldsymbol{v}, t) = \frac{1}{2} X_0 + v_1 Z_0 Z_1 + \sin(v_2 t) X_1
 
-        and the observable :math:`B=I\otimes Y`. There are two variational parameters, :math:`v_1`
+        and the observable :math:`B= Y_1`. There are two variational parameters, :math:`v_1`
```
Comment (dwierichs) [inline pennylane/gradients/pulse_gradient.py:547]:
```suggestion
        and the observable :math:`B=Y_1`. There are two variational parameters, :math:`v_1`
```

**Review by dwierichs — APPROVED** [review]
Thanks for these neat changes, @Qottmann ! :100:
I like all of them! :) 
Just a few minor suggestions/comments, feel free to include/exclude whichever you see fit!

File: pennylane/gradients/pulse_generator_gradient.py (hunk 3 above)
Comment (Qottmann) [reply pennylane/gradients/pulse_generator_gradient.py:625]: Ah you mean even though the dynamical lie algebra has some size, it could be that some coefficients in the decomposition are zero?

File: pennylane/gradients/pulse_gradient.py (hunk 5 above)
Comment (Qottmann) [reply pennylane/gradients/pulse_gradient.py:422]: Oh thanks, no! Missed that detail!

File: pennylane/gradients/pulse_generator_gradient.py (hunk 3 above)
Comment (Qottmann) [reply pennylane/gradients/pulse_generator_gradient.py:625]: updated the sentence

**Review by frederikwilde — CHANGES_REQUESTED** [review]
Besides the mistake in the PSR it looks good to me. Just two nit-picky and potentially incorrect language remarks (I'm not a native speaker). The changes definitely improve the accessibility! 👍

File: pennylane/gradients/pulse_gradient.py (hunk 6 above)
Comment (frederikwilde) [reply pennylane/gradients/pulse_gradient.py:469]: These numbers are dependent on the computer (and its CPU utilization) on which the code is executed, no?

Comment (Qottmann) [reply pennylane/gradients/pulse_gradient.py:469]: yes 💯 ... and the temperature of the computer and the room that computer is in 😄 and how many other things are running etc.

File: pennylane/gradients/pulse_generator_gradient.py [hunk 8]
```diff
@@ -658,14 +662,20 @@ def circuit(params):
         before the parametrized time evolution. Here, the variable :math:`x` is just a
         convenient way to write the modified cost function. Note that its derivative with
         respect to :math:`x` can be computed with the standard two-term parameter-shift
-        rule for Pauli rotation gates.
+        rule for Pauli rotation gates, i.e.
+
+        .. math::
+
+            \frac{\mathrm{d}}{\mathrm{d}x} C_\ell(x)\large|_{x=0} = \frac{1}{2} \left(C_\ell(\pi/2) - C_\ell(-\pi/2)\right)
```
[the remark this answers is not in the document]
Comment (Qottmann) [reply pennylane/gradients/pulse_generator_gradient.py:669]: With the explicit $1/2$ in the definition we need $\pi/2$ such that the overall factor in the exponent is $\pi/4$, or am I standing on the tube? 😅

File: pennylane/gradients/pulse_generator_gradient.py [hunk 9]
```diff
@@ -619,7 +620,10 @@ def circuit(params):
 
             \omega_{k}^{(\ell)} = \frac{1}{2^N}\mathrm{Tr}\left[P_\ell \Omega_{k}\right]
 
-        where :math:`N` is the number of qubits.
+        where :math:`N` is the number of qubits and :math:`\ell = 1, .. , L` the Pauli word index.
+        The number of non-zero Pauli words :math:`L` is typically the dimension of the dynamical Lie algebra
+        (can be lower if coefficients happen to be zero)
+        and at most :math:`4^N-1`.
```
[the remark this answers is not in the document]
Comment (Qottmann) [reply pennylane/gradients/pulse_generator_gradient.py:626]: Would: "dimension of the DLA" make more sense as opposed to "number of dimensions of the DLA"?

[the remark this answers is not in the document]
Comment (Qottmann) [reply pennylane/gradients/pulse_generator_gradient.py:626]: I think adding "equal to" still makes sense :) thanks for that!

File: pennylane/gradients/pulse_generator_gradient.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (frederikwilde) [reply pennylane/gradients/pulse_generator_gradient.py]: You are right, $1/2$ is the correct shift. Sorry about this!
